ARTEMIS-3794 System Property Encryption Support

ryan-highley opened the pull request:
Adds support for standard Java TLS and ActiveMQ Artemis-specific override
encrypted system property values for the key store and trust store
passwords

brusdev commented on tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/remoting/impl/netty/NettyConnectorTest.java:
```diff
@@ -171,6 +173,70 @@ public void bufferReceived(final Object connectionID, final ActiveMQBuffer buffe
+
+      DefaultSensitiveStringCodec codec = new DefaultSensitiveStringCodec();
+
+      System.setProperty(NettyConnector.JAVAX_KEYSTORE_PATH_PROP_NAME, "client-keystore.jks");
+      System.setProperty(NettyConnector.JAVAX_KEYSTORE_PASSWORD_PROP_NAME, PasswordMaskingUtil.wrap(codec.encode("bad password")));
+      System.setProperty(NettyConnector.JAVAX_TRUSTSTORE_PATH_PROP_NAME, "server-ca-truststore.jks");
+      System.setProperty(NettyConnector.JAVAX_TRUSTSTORE_PASSWORD_PROP_NAME, PasswordMaskingUtil.wrap(codec.encode("bad password")));
```
The system properties should be cleared before ending the test because they could affect other tests.

brusdev commented on tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/remoting/impl/netty/NettyConnectorTest.java:
```diff
@@ -334,6 +400,59 @@ public void bufferReceived(final Object connectionID, final ActiveMQBuffer buffe
+
+      DefaultSensitiveStringCodec codec = new DefaultSensitiveStringCodec();
+
+      System.setProperty(NettyConnector.ACTIVEMQ_KEYSTORE_PATH_PROP_NAME, "client-keystore.jks");
+      System.setProperty(NettyConnector.ACTIVEMQ_KEYSTORE_PASSWORD_PROP_NAME, PasswordMaskingUtil.wrap(codec.encode("securepass")));
+      System.setProperty(NettyConnector.ACTIVEMQ_TRUSTSTORE_PATH_PROP_NAME, "server-ca-truststore.jks");
+      System.setProperty(NettyConnector.ACTIVEMQ_TRUSTSTORE_PASSWORD_PROP_NAME, PasswordMaskingUtil.wrap(codec.encode("securepass")));
```
The system properties should be cleared before ending the test because they could affect other tests.

brusdev commented:
I see a test failure due to system properties set by the new test:
https://github.com/apache/activemq-artemis/runs/7171602475?check_suite_focus=true
```
WARN  [io.netty.channel.ChannelInitializer] Failed to initialize a channel. Closing: [id: 0xcf2b8ab7]: java.lang.Exception: Failed to find a store at unknownclient-keystore.jks
	at org.apache.activemq.artemis.core.remoting.impl.ssl.SSLSupport.validateStoreURL(SSLSupport.java:356)
	at org.apache.activemq.artemis.core.remoting.impl.ssl.SSLSupport.loadKeystore(SSLSupport.java:304)
	at org.apache.activemq.artemis.core.remoting.impl.ssl.SSLSupport.loadKeyManagerFactory(SSLSupport.java:332)
	...
```

brusdev commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyConnector.java:
```diff
@@ -584,15 +585,15 @@ public synchronized void start() {
          } else {
             realKeyStorePath = Stream.of(System.getProperty(ACTIVEMQ_KEYSTORE_PATH_PROP_NAME), System.getProperty(JAVAX_KEYSTORE_PATH_PROP_NAME), keyStorePath).map(v -> useDefaultSslContext ? keyStorePath : v).filter(Objects::nonNull).findFirst().orElse(null);
-            realKeyStorePassword = Stream.of(System.getProperty(ACTIVEMQ_KEYSTORE_PASSWORD_PROP_NAME), System.getProperty(JAVAX_KEYSTORE_PASSWORD_PROP_NAME), keyStorePassword).map(v -> useDefaultSslContext ? keyStorePassword : v).filter(Objects::nonNull).findFirst().orElse(null);
+            realKeyStorePassword = processSslPasswordProperty(Stream.of(System.getProperty(ACTIVEMQ_KEYSTORE_PASSWORD_PROP_NAME), System.getProperty(JAVAX_KEYSTORE_PASSWORD_PROP_NAME), keyStorePassword).map(v -> useDefaultSslContext ? keyStorePassword : v).filter(Objects::nonNull).findFirst().orElse(null));
```
The system properties have a grater visibility than any connector properties and multiple connector could rely on the same system properties so it doesn't sound good to me that a connector property affects the use of a system property.

ryan-highley replied:
The stream processing logic is identical to the previous code--I just wrapped the stream result in a method call to handle the possibility of an encrypted system property. However, the encrypted password logic should only be invoked when the value is not the keyStorePassword or trustStorePassword captured with ConfigurationHelper#getPasswordProperty(...) previously as that value has already been decrypted if necessary.

I've tweaked the logic to call processSslPasswordProperty(...) only when the existing stream processing result is one of the system properties, not a query string value.

brusdev replied:
I'm not sure to get the purpose of this PR, why does a user need to set the password-codec in the connector url and the passwords in system properties?

ryan-highley replied:
This PR enables ENC(...) handling for the 'org.apache.activemq.ssl.keyStorePassword', 'o.a.a.ssl.trustStorePassword', 'javax.net.ssl.keyStorePassword', and 'j.n.ssl.trustStorePassword' system property values. The connector URL 'keyStorePassword' and 'trustStorePassword' query string handling is unchanged as these values already are decrypted as needed through the ConfigurationHelper#getPasswordProperty(...) method in lines 401 and 409.

The system properties obviously aren't part of the 'configuration' Map so the ConfigurationHelper#getPasswordProperty(...) method can't be used directly. The processSslPasswordProperty(...) method duplicates the behavior of calling getPasswordProperty(...) using the same ActiveMQDefaultConfiguration.getPropMaskPassword() and .getPropPasswordCodec() used for the connector URL password handling for the 'keyStorePassword' and 'trustStorePassword'.

Hopefully, that makes more sense as to the intent. ARTEMIS-3794 is just a JIRA I saw similar to the AMQP broker connection user and password attribute encryption handling.

brusdev replied:
If the purpose is to override keyStore/trustStore passwords obtained from topology with encrypted system properties maybe we should use system properties to set the mask mode and the password codec too, i.e. 'org.apache.activemq.ssl.passwordCodec', the server and the client could use a different mask mode or password codec.

ryan-highley replied:
That's an interesting thought. I based these updates on the approach handling keyStorePassword and trustStorePassword also applying adequately for their corresponding system property values.

I'm all for providing essential flexibility, but I've also never had to provide an alternate password encoding codec to appease PEN testing or production environment security scans. I'm happy to address that if you see the need for that additional configuration and associated incremental complexity.

Perhaps we can see if anyone files an enhancement request for specifying an alternate mask mode and/or password codec with system properties?

jbertram commented:
@ryan-highley, can you squash and rebase?

ryan-highley commented:
@jbertram @brusdev 

Rebased and squashed these updates.

Any thoughts on adding two additional system properties for the password mask and codec configurations? Could always handle those in a different RFE as well.

jbertram commented:
@ryan-highley, I'm not really in favor of a "useMaskedPassword" property. We've moved away from this style of configuration in recent years and consolidated on the `ENC()` syntax which can be automatically detected.

ryan-highley commented:
> @ryan-highley, I'm not really in favor of a "useMaskedPassword" property. We've deprecated this style of configuration in recent years and consolidated on the `ENC()` syntax which can be automatically detected.

Agree wholeheartedly!

The parameter-based approach is documented as being deprecated, and I'm perfectly happy to remove its support with the system properties here. I included the system property solely for "principle of least surprise" since elsewhere the "useMaskedPassword" configurations do still work for backwards compatibility.

I'll remove the system property and gladly clean up the associated extraneous unit tests!

jbertram commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyConnector.java:
```diff
@@ -755,6 +771,23 @@ public void initChannel(Channel channel) throws Exception {
       logger.debug("Started {} Netty Connector version {} to {}:{}", connectorType, TransportConstants.NETTY_VERSION, host, port);
    }
 
+   private String processSslPasswordProperty(String password, String codecClass) {
```
Instead of using `ConfigurationHelper` you could simply use `PasswordMaskingUtil` (which is what `ConfigurationHelper` does eventually), e.g.:
```java
   private String resolveMask(String password, String codecClass) {
      try {
         return PasswordMaskingUtil.resolveMask(password, codecClass);
      } catch (Exception e) {
         throw new RuntimeException(e);
      }
   }
```

ryan-highley replied:
Oh, yes. Much better! I wanted to reuse the `ConfigurationHelper` password logic since the extra Boolean logic was already handled there. Without that `useMaskedPassword` that reasoning and need goes right out the window.

Updated and pushed.

jbertram commented:
I'm ready to merge this except for the comment I've made. Let me know what you think.

jbertram commented:
@ryan-highley, thanks for the contribution and also thanks for being willing to discuss and accept feedback. It's great working with you!
